lua/entities/gmod_tardis/modules/sh_chameleon.lua

```diff
@@ -0,0 +1,277 @@
+        local pos2 = self:GetPos() + self:GetUp() * (modelmaxs.z - (modelheight * new_percent))
+
+        self:SetData("phase-highPos", pos)
+        self:SetData("phase-pos", pos2)
+    end)
+
+    ENT:AddHook("ShouldDrawPhaseAnimation", "chameleon", function(self)
+        if self:GetData("chameleon_animation",false) then
+            return true
+        end
+    end)
+
+    ENT:AddHook("ExteriorChanged", "scale", function(self)
+        self:DisableMatrix("RenderMultiply")
+    end)
+end
+
+function ENT:ChangeExteriorMetadata(id)
+    if SERVER then
+        self:SendMessage("exterior_metadata_update", {id})
+    end
+
+    local original_md = self.metadata.ExteriorOriginal
+    if original_md == nil then
+        self.metadata.ExteriorOriginal = self.metadata.Exterior
+    end
+
+    local ext_md = (id == nil and original_md) or TARDIS:CreateExteriorMetadata(id)
+
+    local oldvortex = self.metadata.Exterior.Parts.vortex
+    if oldvortex then
+        ext_md.Parts.vortex = TARDIS:CopyTable(oldvortex)
+    end
+
+    self.metadata.Exterior = ext_md
+    self.interior.metadata.Exterior = ext_md
+
+    return ext_md
+end
+
+if CLIENT then
+    ENT:OnMessage("exterior_metadata_update", function(self,data,ply)
+        self:ChangeExteriorMetadata(data[1])
+    end)
+end
+
+function ENT:ChangeExterior(id, animate)
+    if CLIENT then
+        self:SendMessage("chameleon_change_exterior", {id, animate})
+        return
+    end
+
+    if self:CallCommonHook("CanChangeExterior", id) == false then
+        self:SetData("chameleon_selected_exterior", id, true)
+        return
+    end
+    self:SetData("chameleon_selected_exterior", nil, true)
+
+    if not IsValid(self.interior) then
+        return
+    end
+
+    local ext_md = self:ChangeExteriorMetadata(id)
+
+    if animate then
+        self:SendMessage("chameleon_exterior_animation")
+    end
+
+    local anim_time = self.metadata.Exterior.ChameleonAnimTime
+    local delay = (animate and anim_time / 2) or 0
+
+    if animate then
+        self:UpdateShadow()
+        self:SetData("chameleon_changing", true, true)
+
+        self:Timer("chameleon_changing_reset", anim_time, function()
+            self:SetData("chameleon_changing", false, true)
+            self:UpdateShadow()
+        end)
+    end
+
+    self:Timer("chameleon_change", delay, function()
+        self:SetModel(ext_md.Model)
+        self:PhysicsInit(SOLID_VPHYSICS)
+        self:SetMoveType(MOVETYPE_VPHYSICS)
+        self:SetSolid(SOLID_VPHYSICS)
+        self.phys = self:GetPhysicsObject()
+
+        if (IsValid(self.phys)) then
+            self.phys:SetMass(ext_md.Mass)
+            self.phys:EnableGravity(not self:GetData("float"))
+            self.phys:EnableMotion(not self:GetPhyslock())
+            self.phys:Wake()
+        end
+
+        local portal = self.interior.portals.exterior
+        local portal_md = ext_md.Portal
+
+        if not IsValid(portal) then return end
```

Comment:
Is this likely to happen - what is the behaviour if there's no portal?

Reply:
This shouldn't happen - at least for now we don't support exteriors without a portal, the portal is in the game

But since we address the portal later, we must at least make sure it exists
General safeguard